From 92263db4ce9c4832f51aeda70fd5361873e8095a Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 3 Jun 2023 16:14:46 +0200
Subject: [PATCH 0817/1013] evl/timer: fix SMP race in overrun handling

evl_get_timer_overruns() might attempt to dequeue an unqueued timer
when handling an overrun, if the caller and the timer run on distinct
CPUs.

Fix this by detecting the pattern illustrated below, in which case
evl_get_timer_overruns() may leave the requeuing operation to the core
clock tick handler.

T: a thread waiting for wakeup events on a periodic timeline.
t: the periodic timer.

CPU0                        CPU1
----                        ----

                            T: <wait_for_next_period>
do_clock_tick()
    lock(t->base)
    dequeue(t)
    unlock(t->base)
       t->handler()
    lock(t->base)
    enqueue(t)
    unlock(t->base)
    remote-wakeup(T)

<OVERRUN>

do_clock_tick()
    lock(t->base)
    dequeue(t)
    unlock(t->base)
                            T: evl_get_timer_overruns()
			            lock(t->base)
			            dequeue(t)
				    <!> double removal

See https://lore.kernel.org/xenomai/874jnoyghn.fsf@xenomai.org/T/

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/evl/timer.h |  4 ++--
 kernel/evl/clock.c  |  2 +-
 kernel/evl/timer.c  | 52 +++++++++++++++++++++++++++++++++++++++------
 3 files changed, 48 insertions(+), 10 deletions(-)

diff --git a/include/evl/timer.h b/include/evl/timer.h
index 0f5616b26e38..c5ce4a0fff40 100644
--- a/include/evl/timer.h
+++ b/include/evl/timer.h
@@ -179,7 +179,7 @@ struct evl_timer {
 	int status;
 	ktime_t interval;	/* 0 == oneshot */
 	ktime_t start_date;
-	u64 pexpect_ticks;	/* periodic release date */
+	u64 consumed_ticks;	/* advanced by evl_get_timer_overruns() */
 	u64 periodic_ticks;
 #ifdef CONFIG_SMP
 	struct evl_rq *rq;
@@ -263,7 +263,7 @@ static inline
 ktime_t evl_get_timer_next_date(struct evl_timer *timer)
 {
 	return ktime_add_ns(timer->start_date,
-			timer->pexpect_ticks * ktime_to_ns(timer->interval));
+			timer->consumed_ticks * ktime_to_ns(timer->interval));
 }
 
 void __evl_init_timer(struct evl_timer *timer,
diff --git a/kernel/evl/clock.c b/kernel/evl/clock.c
index ca5422f2de3f..e72db15cc6ca 100644
--- a/kernel/evl/clock.c
+++ b/kernel/evl/clock.c
@@ -85,7 +85,7 @@ static void adjust_timer(struct evl_clock *clock,
 		 */
 		div = ktime_divns(-diff, ktime_to_ns(period));
 		timer->periodic_ticks -= div;
-		timer->pexpect_ticks -= div;
+		timer->consumed_ticks -= div;
 		evl_update_timer_date(timer);
 	}
 
diff --git a/kernel/evl/timer.c b/kernel/evl/timer.c
index c609676eae5b..a53a53eb3e6f 100644
--- a/kernel/evl/timer.c
+++ b/kernel/evl/timer.c
@@ -160,7 +160,7 @@ void evl_start_timer(struct evl_timer *timer,
 	if (!timeout_infinite(interval)) {
 		timer->interval = interval;
 		timer->start_date = value;
-		timer->pexpect_ticks = 0;
+		timer->consumed_ticks = 0;
 		timer->periodic_ticks = 0;
 		timer->status |= EVL_TIMER_PERIODIC;
 	}
@@ -415,21 +415,59 @@ unsigned long evl_get_timer_overruns(struct evl_timer *timer)
 	struct evl_tqueue *tq;
 	ktime_t now, delta;
 
-	now = evl_read_clock(timer->clock);
+	if (EVL_WARN_ON_ONCE(CORE, !evl_timer_is_periodic(timer)))
+		return 0;
+
 	base = lock_timer_base(timer, &flags);
+	now = evl_read_clock(timer->clock);
 
+	/*
+	 * Measure the lateness with respect to the expected expiry
+	 * date in the timeline for this timer. This ideal expiry date
+	 * would be the start date of the timer, plus the interval
+	 * duration times the number of ticks which were consumed so
+	 * far. Any delay longer than the periodic interval denotes an
+	 * overrun condition since we missed at least one tick.
+	 */
 	delta = ktime_sub(now, evl_get_timer_next_date(timer));
 	if (likely(delta < timer->interval))
 		goto done;
 
 	overruns = ktime_divns(delta, ktime_to_ns(timer->interval));
-	timer->pexpect_ticks += overruns;
+
+	/*
+	 * Compensate for the current lateness as the timer progresses
+	 * in the timeline, consuming all missed ticks in one gulp to
+	 * catch up.
+	 */
+	timer->consumed_ticks += overruns;
+
+	/*
+	 * We are about to re-enqueue the timer for its next expiry
+	 * date. However, we refrain from doing so if either:
+	 *
+	 * - the timer was deactivated, in which case there is no
+	 * point in planning for further expiry.
+	 *
+	 * - the timer is still running but not queued, which means
+	 * that do_clock_tick() dequeued it then released the base
+	 * lock on a remote CPU for running its handler. Since the
+	 * timer is running and we hold that lock at the moment,
+	 * do_clock_tick() is certainly going to either advance the
+	 * expiry as required or shut the timer down after we drop the
+	 * base lock. IOW, we may leave reprogramming to the core tick
+	 * handler.
+	 */
 	if (!evl_timer_is_running(timer))
 		goto done;
 
-	EVL_WARN_ON_ONCE(CORE, (timer->status &
-			(EVL_TIMER_DEQUEUED|EVL_TIMER_PERIODIC))
-			!= EVL_TIMER_PERIODIC);
+	if (timer->status & EVL_TIMER_DEQUEUED)
+		goto done;
+
+	/*
+	 * Re-enqueue the periodic timer so that it ticks at the next
+	 * interval boundary.
+	 */
 	tq = &base->q;
 	evl_dequeue_timer(timer, tq);
 	while (evl_tdate(timer) < now) {
@@ -439,7 +477,7 @@ unsigned long evl_get_timer_overruns(struct evl_timer *timer)
 
 	program_timer(timer, tq);
 done:
-	timer->pexpect_ticks++;
+	timer->consumed_ticks++;
 
 	unlock_timer_base(base, flags);
 
-- 
2.30.2

